src/lib.rs

```diff
@@ -1074,84 +1106,200 @@ impl<
     }
 
     /// Add a viewing key to the keystore's key set.
+    ///
+    /// Since this key was not generated by this keystore, it may have already been used and thus may
+    /// have viewable records. The keystore will start a scan of the ledger in the background to
+    /// find records viewable by this key. The scan will start from the event specified by
+    /// `scan_from`.
     pub fn add_viewing_account<'l>(
         &'l mut self,
         viewing_key: ViewerKeyPair,
         description: String,
+        scan_from: EventIndex,
     ) -> std::pin::Pin<Box<dyn SendFuture<'a, Result<(), KeystoreError<L>>> + 'l>>
     where
         'a: 'l,
     {
         Box::pin(async move {
-            self.write()
+            let (viewing_key, events) = self
+                .write()
                 .await
-                .update(|KeystoreSharedState { state, model, .. }| {
-                    state.add_viewing_account(model, Some(viewing_key), description)
-                })
+                .update(
+                    |KeystoreSharedState {
+                         state,
+                         model,
+                         pending_viewing_key_scans,
+                         ..
+                     }| async move {
+                        let (viewing_key, events) = state
+                            .add_viewing_account(
+                                model,
+                                Some(viewing_key),
+                                description,
+                                Some(scan_from),
+                            )
+                            .await?;
+                        // Register the key scan in `pending_viewing_key_scans` so that
+                        // `await_viewing_key_scan` will work.
+                        pending_viewing_key_scans.insert(viewing_key.pub_key(), vec![]);
+                        Ok((viewing_key, events))
+                    },
+                )
                 .await?;
+
+            if let Some(events) = events {
+                // Start a background task to scan for records viewable by the new key.
+                self.spawn_viewing_key_scan(viewing_key.pub_key(), events)
+                    .await;
+            }
+
```

Comment:
Is it possible to remove the duplicated code at least per key type?

Reply:
I tried using two enums `KeyType` and `PubKeyType` to avoid duplications for different `KeyPair` and `PubKey` types, respectively. `KeyType` simplified the relevant code a lot, but `PubkeyType` resulted in much more code which didn't seem that helpful, so I kept `KeyType` but not `PubkeyType`. I'll see if I can simplify this part though.

Reply:
Oops, I just realized that the inputs of these functions are _KeyPair_, rather than _PubKey_, so I can definitely simplify this using the `KeyType` enum.